analytical_engine/apps/simple_path/all_simple_paths.h

```diff
@@ -0,0 +1,299 @@
+    std::vector<typename fragment_t::oid_t> data;
+    std::set<vid_t> qvisit;
+    std::vector<vid_t> q;
+    vid_t source_gid;
+    frag.Oid2Gid(ctx.source_id, source_gid);
+    qvisit.insert(source_gid);
+    q.push_back(source_gid);
+    int index = find_edge_map_index(ctx, source_gid);
+    Pint_Result(index, 0, q, qvisit, data, frag, ctx);
+    std::vector<size_t> shape{static_cast<size_t>(ctx.path_num),
+                              static_cast<size_t>(ctx.cutoff + 1)};
+    if (ctx.path_num == 0) {
+      data.push_back(1);
+      shape.pop_back();
+      shape[0] = 1;
+    }
+    VLOG(0) << "path_num: " << ctx.path_num << std::endl;
+    ctx.assign(data, shape);
+  }
+  void Pint_Result(int from, int depth, std::vector<vid_t>& q,
+                   std::set<vid_t>& qvisit,
+                   std::vector<typename fragment_t::oid_t>& data,
+                   const fragment_t& frag, context_t& ctx) {
+    if (depth == (ctx.cutoff - 1)) {
+      typename std::set<vid_t>::iterator it;
+      for (it = ctx.targets.begin(); it != ctx.targets.end(); it++) {
+        auto t = *it;
+        if (qvisit.count(t) == 1) {
+          continue;
+        }
+        int to = find_edge_map_index(ctx, t);
+        if (std::find(ctx.edge_map[from].begin(), ctx.edge_map[from].end(),
+                      to) != ctx.edge_map[from].end()) {
+          int len_counter = 0;
+          for (auto gid : q) {
+            data.push_back(frag.Gid2Oid(gid));
+            len_counter++;
+          }
+          data.push_back(frag.Gid2Oid(t));
+          len_counter++;
+          // fill path to cutoff with -1
+          while (len_counter != ctx.cutoff + 1) {
+            data.push_back(-1);
+            len_counter++;
+          }
+          ctx.path_num++;
+        }
+      }
+      return;
+    }
+
+    for (uint64_t t = 0; t < ctx.edge_map[from].size(); t++) {
+      int to = ctx.edge_map[from][t];
+      vid_t gid = index2gid(ctx, to);
+      if (qvisit.count(gid) == 1) {
+        continue;
+      }
+      qvisit.insert(gid);
+      if (ctx.targets.count(gid)) {
+        int len_counter = 0;
+        for (auto v : q) {
+          data.push_back(frag.Gid2Oid(v));
+          len_counter++;
+        }
+        data.push_back(frag.Gid2Oid(gid));
+        len_counter++;
+        // fill path length to cutoff with -1
+        while (len_counter != ctx.cutoff + 1) {
+          data.push_back(-1);
+          len_counter++;
+        }
+        ctx.path_num++;
+      }
+      q.push_back(gid);
+      Pint_Result(to, depth + 1, q, qvisit, data, frag, ctx);
+      q.pop_back();
+      qvisit.erase(gid);
+    }
+  }
+
+  int find_edge_map_index(context_t& ctx, vid_t gid) {
+    int fid = static_cast<int>(gid >> ctx.fid_offset);
+    int lid = static_cast<int>(gid & ctx.id_mask);
+    int ret = 0;
+    for (int i = 0; i < fid; i++) {
+      ret += ctx.frag_vertex_num[i];
+    }
+    return ret + lid;
+  }
+
+  vid_t index2gid(context_t& ctx, int index) {
```

Comment:
Why we have index2gid here since context already implement one?

Reply:
context do not need context_t& ctx, just distinguish it from context.h.